@xen-orchestra/backups-cli/index.js

```
@@ -0,0 +1,354 @@
+  const children = chain.slice(0, -1).reverse()
+
+  console.warn('Unused parents of VHD', child)
+  chain
+    .slice(1)
+    .reverse()
+    .forEach(parent => {
+      console.warn(' ', parent)
+    })
+  force && console.warn('  merging…')
+  console.warn('')
+  if (force) {
+    // `mergeVhd` does not work with a stream, either
+    // - makes it accept a stream
+    // - or create synthetic VHD which is not a stream
+    throw new Error('FIXME')
+    // await mergeVhd(
+    //   handler,
+    //   parent,
+    //   handler,
+    //   children.length === 1
+    //     ? child
+    //     : await createSyntheticStream(handler, children)
+    // )
+  }
+
+  await Promise.all([
+    fs.rename(parent, child),
+    asyncMap(children.slice(0, -1), child => {
+      console.warn('Unused VHD', child)
+      force && console.warn('  deleting…')
+      console.warn('')
+      return force && handler.unlink(child)
+    }),
+  ])
+}
+
+const listVhds = pipe([
+  vmDir => vmDir + '/vdis',
+  readDir,
+  asyncMap(readDir),
+  flatten,
+  asyncMap(readDir),
+  flatten,
+  filter(_ => _.endsWith('.vhd')),
+])
+
+// TODO: add lock on VM dir
+async function handleVm(vmDir) {
+  const vhds = new Set()
+  const vhdParents = { __proto__: null }
+  const vhdChildren = { __proto__: null }
+
+  // remove broken VHDs
+  await asyncMap(await listVhds(vmDir), async path => {
+    try {
+      const vhd = new Vhd(handler, path)
+      await vhd.readHeaderAndFooter()
+      vhds.add(path)
+      if (vhd.footer.diskType === DISK_TYPE_DIFFERENCING) {
+        const parent = resolve(dirname(path), vhd.header.parentUnicodeName)
+        vhdParents[path] = parent
+        if (parent in vhdChildren) {
+          const error = new Error(
+            'this script does not support multiple VHD children'
+          )
+          error.parent = parent
+          error.child1 = vhdChildren[parent]
+          error.child2 = path
+          throw error // should we throw?
```

When/why does it happen?

---

It should not happen, but I prefer failing hard on this, instead of silently screwing things.